### PR TITLE
Improved new-package script.

### PR DESCRIPTION
**Pull request by flying-sheep**
Basically rewrote it future-based.

### CODE REVIEW

File: scripts/new-package.ts [hunk 1]
```diff
@@ -1,47 +1,111 @@
...
+        "files": [
+            "index.d.ts",
+            `${newPackageName}-tests.ts`
+        ]
+    }
+
+    return JSON.stringify(config, undefined, 4)
+}
 
 write("tslint.json", '{ "extends": "../tslint.json" }');
```
Comment (andy-ms) [inline scripts/new-package.ts]: This belongs in `const files`.

File: scripts/new-package.ts [hunk 2]
```diff
@@ -1,47 +1,111 @@
...
+            get(`http://registry.npmjs.org/${newPackageName}`, (res) => {
+                if (res.statusCode !== 200) return reject(res.statusCode)
+                let rawData = ""
+                res.on("data", chunk => rawData += chunk)
+                res.on("end", () => resolve(JSON.parse(rawData)))
+            }).on("error", e => reject(e))
+        })
+
+        version = pkg["dist-tags"].latest
+        homepage = pkg.versions[version].homepage
```
Comment (andy-ms) [inline scripts/new-package.ts]: This may be undefined. (Tested with `npm run new-package -- foo`.)

File: scripts/new-package.ts [hunk 3]
```diff
@@ -1,47 +1,111 @@
...
+        const pkg: any = await new Promise((resolve, reject) => {
+            get(`http://registry.npmjs.org/${newPackageName}`, (res) => {
+                if (res.statusCode !== 200) return reject(res.statusCode)
+                let rawData = ""
+                res.on("data", chunk => rawData += chunk)
+                res.on("end", () => resolve(JSON.parse(rawData)))
+            }).on("error", e => reject(e))
+        })
+
+        version = pkg["dist-tags"].latest
```
Comment (andy-ms) [inline scripts/new-package.ts]: Extract just the major/minor; we don't want the patch version.

**Review by ghost — COMMENTED** [review]
Thanks!

**Comment by flying-sheep** [comment]
all done. i wonder why there’s no package.json interface yet. people could sometimes want to do `const foo: NPMPackage = JSON.parse('package.json')`

File: scripts/new-package.ts [hunk 4]
```diff
@@ -1,47 +1,124 @@
...
+    }
+}
+
+async function getIndex() {
+    let version = "x.x"
+    let project = "https://github.com/baz/foo (Does not have to be to GitHub, but prefer linking to a source code repository rather than to a project website.)"
+    try {
+        const pkg = await new Promise<Registry>((resolve, reject) => {
+            get(`http://registry.npmjs.org/${newPackageName}`, (res) => {
+                if (res.statusCode !== 200) return reject(res.statusCode)
```
Comment (andy-ms) [inline scripts/new-package.ts]: Should wrap this in an error; the `catch` extracts `stack`, which is undefined for a number.

**Comment by flying-sheep** [comment]
a bit sad that i had to define a ghetto `fetch` (`loadString`) and node APIs aren’t promisified.

i really like being able to use `fetch` and `mz` instead of `http` and `fs`

**Comment by andy-ms** [comment]
It's not too late to add those. People will need to `npm install` anyway to get tslint.

**Comment by flying-sheep** [comment]
if you think this is less maintainable the way it is, i will, but since all other scripts just use node builtins i figured i wouldn’t deviate.

the reason i did it like this in the first place is that it was just hard to end up with the control flow i envisioned with a callback-based API (`Promise.all` is ugly to emulate with callbacks)

**Comment by andy-ms** [comment]
OK. This needs a compile one more time (`scripts/new-package.js` is checked in) and then I'll merge.

**Comment by flying-sheep** [comment]
oh, i’m so sorry i forgot to compile it this time!

**Pull request merged**